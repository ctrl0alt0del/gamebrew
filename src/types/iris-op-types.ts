export enum IrisKind {
  Value,
  Stream,
  Action,
  Lens,
  Listen,
  Dispatch,
}

export type IrisOpKind = IrisKind.Listen | IrisKind.Dispatch;

export type IrisValue<_T> = {
  kind: IrisKind.Value;
  id: string;
};

export type Stream<T> = {
  kind: IrisKind.Stream;
  id: string;
  value: IrisValue<T>;
};

export interface BaseIrisOp<OpType extends IrisKind = IrisKind> {
  kind: OpType;
}

export interface IrisListenOp<T> extends BaseIrisOp<IrisKind.Listen> {
  stream: Stream<T>;
  ops: BaseIrisOp[];
}

const IrisKindValues = Object.values(IrisKind);

export const isIrisOp = (op: any): op is BaseIrisOp =>
  op && IrisKindValues.includes(op.kind);

export const isIrisListenOp = (op: BaseIrisOp<any>): op is IrisListenOp<any> =>
  op.kind === IrisKind.Listen;
