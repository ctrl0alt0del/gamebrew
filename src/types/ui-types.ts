type KeyMetaType = string | number | symbol;
export interface IUIElement<
  _Tag extends KeyMetaType = KeyMetaType,
  Props = any
> {
  readonly _tag: _Tag;
  key: string;
  props: Props;
  children?: (IUIElement | string)[];
}

export type UIElementStandard<PropsSet extends { [k: KeyMetaType]: unknown }> =
  {
    [key in keyof PropsSet]: IUIElement<key, PropsSet[key]>;
  }[keyof PropsSet];

type PropsOfUIElementStandard<
  T extends UIElementStandard<any>,
  _Tag extends string
> = T extends UIElementStandard<{
  [k in _Tag]: infer Props;
}>
  ? Props
  : never;

type IrisPropsSet = {
  root: {};
  text: {
    text: string;
  };
  pressable: {
    onPress: () => void;
  };
  view: {};
};
export type IrisElementTags = keyof IrisPropsSet;
export type IrisElement = UIElementStandard<IrisPropsSet>;
export type IrisElementInstance<Tag extends IrisElementTags> = IUIElement<
  Tag,
  IrisElementProps<Tag>
>;
export type IrisElementProps<Tag extends IrisElementTags> =
  PropsOfUIElementStandard<IrisElement, Tag>;

export type RootElement = IUIElement<'root', {}>;

const availableIrisTags = ['text', 'pressable', 'root'] as const;

export const isIrisElementTag = (tag: KeyMetaType): tag is IrisElementTags =>
  availableIrisTags.includes(tag as any);

export const isIrisElement = (element: IUIElement): element is IrisElement =>
  isIrisElementTag(element._tag);
