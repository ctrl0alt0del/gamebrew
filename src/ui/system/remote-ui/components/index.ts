export * from './iris-text';
export * from './iris-view';
