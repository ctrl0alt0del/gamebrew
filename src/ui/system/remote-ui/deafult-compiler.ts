import { factoryIrisCompiler } from './factory-compiler';

export const { compile, useIrisCompile } = factoryIrisCompiler();
