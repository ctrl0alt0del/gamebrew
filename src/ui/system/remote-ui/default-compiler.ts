import { factoryIrisCompiler } from './factory-compiler';
import { reactToIrisTree } from './iris-remote-tree';

export const { compile, useIrisCompile } = factoryIrisCompiler({
  toIrisTree: reactToIrisTree,
});
