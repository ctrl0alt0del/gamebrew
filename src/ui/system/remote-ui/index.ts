export * from './components';
export * from './default-compiler';
export * from './hooks';
export * from './iris-remote-tree';
