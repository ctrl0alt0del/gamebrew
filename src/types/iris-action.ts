import type { IrisActionsKind } from '@/core/iris-actions-lib/iris-action-def';

import type { BaseIrisOp } from './iris-op-types';
import { IrisKind } from './iris-op-types';

export interface IrisActionOp<P extends Array<any> = any[]>
  extends BaseIrisOp<IrisKind.Action> {
  action: IrisActionsKind;
  args: P;
}

export interface IrisDispatchOp extends BaseIrisOp<IrisKind.Dispatch> {
  actions: IrisActionOp[];
}

export const IrisActions = {
  create: <T extends any[]>(
    action: IrisActionsKind,
    ...args: T
  ): IrisActionOp<T> => ({
    kind: IrisKind.Action,
    action,
    args,
  }),
};

export const IrisDispatch = {
  create: (...actions: IrisActionOp[]): IrisDispatchOp => ({
    kind: IrisKind.Dispatch,
    actions,
  }),
  is: (op: any): op is IrisDispatchOp => op.kind === IrisKind.Dispatch,
};
