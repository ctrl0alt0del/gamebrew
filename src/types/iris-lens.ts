import type { BaseIrisOp } from './iris-op-types';
import { IrisKind } from './iris-op-types';

export enum IrisLensKind {
  Prop,
  Index,
  ItemByProp,
  Identity,
  Composed,
}

export interface BaseIrisLensOp<K extends IrisLensKind, _A, _B>
  extends BaseIrisOp<IrisKind.Lens> {
  lensKind: K;
}

export type IrisLensOp<A, B> = BaseIrisLensOp<IrisLensKind, A, B>;

export interface IrisPropLensOp<A extends Record<string, any>, B>
  extends BaseIrisLensOp<IrisLensKind.Prop, A, B> {
  prop: {
    [K in keyof A]: A[K] extends B ? K : never;
  }[keyof A];
}

export interface IrisIndexLensOp<A extends Array<B>, B>
  extends BaseIrisLensOp<IrisLensKind.Index, A, B> {
  index: number;
}

export interface IrisItemByPropLensOp<A extends Array<B>, B>
  extends BaseIrisLensOp<IrisLensKind.ItemByProp, A, B> {
  prop: keyof B;
}

export interface IrisIdentityLensOp<A>
  extends BaseIrisLensOp<IrisLensKind.Identity, A, A> {}

export interface IrisComposedLensOp<A, B, C>
  extends BaseIrisLensOp<IrisLensKind.Composed, A, C> {
  outer: BaseIrisLensOp<any, A, B>;
  inner: BaseIrisLensOp<any, B, C>;
}

export const IrisLens = {
  prop: <A extends Record<string, any>, B>(
    prop: keyof A
  ): IrisPropLensOp<A, B> => ({
    kind: IrisKind.Lens,
    lensKind: IrisLensKind.Prop,
    prop: prop as any,
  }),
  index: <A extends Array<B>, B>(index: number): IrisIndexLensOp<A, B> => ({
    kind: IrisKind.Lens,
    lensKind: IrisLensKind.Index,
    index,
  }),
  itemByProp: <A extends Array<B>, B>(
    prop: keyof B
  ): IrisItemByPropLensOp<A, B> => ({
    kind: IrisKind.Lens,
    lensKind: IrisLensKind.ItemByProp,
    prop: prop as any,
  }),
  identity: <A>(): IrisIdentityLensOp<A> => ({
    kind: IrisKind.Lens,
    lensKind: IrisLensKind.Identity,
  }),
  compose: <A, B, C>(
    outer: BaseIrisLensOp<any, A, B>,
    inner: BaseIrisLensOp<any, B, C>
  ): IrisComposedLensOp<A, B, C> => ({
    kind: IrisKind.Lens,
    lensKind: IrisLensKind.Composed,
    outer,
    inner,
  }),
};
