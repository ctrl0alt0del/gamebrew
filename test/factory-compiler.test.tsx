import { isIrisOp } from '@/types/iris-op-types';
import { compile, IrisText, IrisView } from '@/ui/system/remote-ui';

describe('factory-compiler', () => {
  it('should compile to array of IrisOp', () => {
    const compiled = compile(
      <IrisView>
        <IrisText text="Hello, World" />
      </IrisView>
    );
    expect(compiled).toBeInstanceOf(Array);
    expect(compiled).toHaveLength(2);
    const [view, text] = compiled;
    expect(isIrisOp(view)).toBeTruthy();
    expect(isIrisOp(text)).toBeTruthy();
  });
});
