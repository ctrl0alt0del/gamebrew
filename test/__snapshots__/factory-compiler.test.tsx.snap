// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`factory-compiler should compile 1`] = `undefined`;
